store/database.go

```diff
@@ -0,0 +1,78 @@
+package store
+
+import (
+	"io"
+)
+
+// VersionedReader extends the Reader interface by making all reads versioned.
+type VersionedReader interface {
+	Has(storeKey string, version uint64, key []byte) (bool, error)
+	Get(storeKey string, version uint64, key []byte) ([]byte, error)
+}
+
+// Reader wraps the Has and Get method of a backing data store.
+type Reader interface {
+	// Has retrieves if a key is present in the key-value data store.
+	//
+	// Note: <key> is safe to modify and read after calling Has.
+	Has(storeKey string, key []byte) (bool, error)
```

Comment:
So this is a departure from the `PrefixStore` pattern, which I think is pretty good.  Otherwise e.g. x/bank would be passing "bank" into every Get, Set and Has.  I'm not really sure the the notion of StoreKey belongs in the interface at all.

Reply:
Yes, but the storeKey is provided to the root multi-store which is then passed to the SS database. The SS is a single storage backend which prefixes key/value tuples with the store key, so we need to know that when performing CRUD or iteration ops.